[Concept] Exercise - Switch and Enum (#694)
exercism/cpp: exercises/concept/troll-the-trolls/.docs/instructions.md

# Instructions

Your friend Helma made a small online game that rapidly gained popularity.
It is called _HellMath_.
The small community attracted some trolls who make the game and the forums pretty unpleasant.

> Helma has asked you to work on a new permission system to separate "the wheat from the chaff". 

Reviewer: I'm not sure if phrases like that are easy to understand for somebody who does not speak English fluently. I'd suggest rephrasing the sentence and making it simpler.

The forum supports three different actions:

- Read
- Write
- Remove

There are four types of accounts, each with different default permissions:

- Guests: can read posts.
- Users and Trolls: can read and write posts.
- Moderators: can read, write and remove posts, they have all the permissions.

Helma has noticed that it is no use to ban troll accounts.
Her strategy is to give them the illusion that their time is "well invested", but their posts are only shown to other trolls.

> For anything that requires a priority, trolls get the least places.

Reviewer: "get the least places" sounds strange. let's hear what @BethanyG has to say.

When they enter a game, the pool of available players is also limited to other trolls.

## 1. Set up user types and permissions.

> First, define an `AccountStatus` enumeration to represent the three account types: `troll`, `guest`, `user`, and `mod`.

Reviewer: three account types -> four account types

> Next, define an `Action` enumeration to represent the three permission types: `read`, `write`, `remove`.

Reviewer: read, write, remove -> read, write, and remove
or remove the "and" from the previous line.

> ## 2. Promote trolls to other trolls.

Reviewer: I'd suggest adding "only" to that sentence.

Every post on the forums saves the `AccountStatus` of the poster in its metadata. 
Make sure that the trolls can only interact with each other.

Helma needs a `display_post` function, that gets two arguments of `AccountStatus` and returns a `bool`.
The first argument is the status of the poster, the second one is the status of the viewer.

```cpp
hellmath::display_post(hellmath::AccountStatus::troll, hellmath::AccountStatus::user);
// => false
hellmath::display_post(hellmath::AccountStatus::mod, hellmath::AccountStatus::guest);
// => true
```

## 3. Check permissions for different users.

Helma needs a way to check if a certain action is allowed for a user.

Please implement a `permission_check` function, that takes an `Action` as a first argument and an `AccountStatus` to check against.
It should return a `bool` according to the listing above.

```cpp
hellmath::permission_check(hellmath::Action::remove, hellmath::AccountStatus::guest);
// => false
hellmath::permission_check(hellmath::Action::write, hellmath::AccountStatus::mod);
// => true
```

> ## 4. Grant game access and pair players

Reviewer: either end all headings with a period or none of them

To keep the actual players in the game accountable for their actions, _Hellmath_ denies access to guest users.
As mentioned above, Helma wants trolls to troll other trolls.
Game connections between other users are unrestricted.

> Implement the `valid_player_combination` function that checks if the group can play together.

Reviewer: the group -> two players

> The function has two arguments of type `AccountStatus` and returns a `bool`

Reviewer: I'd say a function *has* parameters or *gets* arguments.

```cpp
hellmath::valid_player_combination(hellmath::AccountStatus::guest, hellmath::AccountStatus::mod);
// => false
hellmath::valid_player_combination(hellmath::AccountStatus::troll, hellmath::AccountStatus::troll);
// => true
```

## 5. Build priority queuing

With the stark growth of the game, Helma has to distribute computing power and bandwidth among users.
To handle emergencies, moderators have the highest priority.

> Guest will be queued behind normal users and trolls get the lowest rank.

Reviewer: rank or priority? be consistent

// => false
display_post(AccountStatus::mod, AccountStatus::guest);
// => true
```

## 3. Check permissions for different users.

Helma needs a way to check if a certain action is allowed for a user.

Please implement a `permission_check` function, that takes an `Action` as a first argument and an `AccountStatus` to check against.
It should return a `bool` according to the listing above.

```cpp
permission_check(Action::remove, AccountStatus::guest);
// => false
permission_check(Action::write, AccountStatus::mod);
// => true
```

## 4. Grant game access and pair players.

To keep the actual players in the game accountable for their actions, _Hellmath_ denies access to guest users.
As mentioned above, Helma wants trolls to troll other trolls.
Game connections between other users are unrestricted.

Implement the `valid_player_combination` function that checks if two players can join the same game.
The function has two parameters of type `AccountStatus` and returns a `bool`.

```cpp
valid_player_combination(AccountStatus::guest, AccountStatus::mod);
// => false
valid_player_combination(AccountStatus::troll, AccountStatus::troll);
// => true
```

## 5. Build priority queuing.

> With the stark growth of the game, Helma has to distribute computing power and bandwidth among users.

Reviewer 2: ```suggestion
With the massive growth of the game and the forums, Helma now has to distribute computing power and bandwidth among users.
```

> To handle emergencies, moderators have the highest priority.

Reviewer 2: ```suggestion
To handle emergencies, moderators are given the highest priority.
```

> Guest will be queued behind normal users and trolls get sorted behind everyone else.

Reviewer 2: ```suggestion
Guests are queued behind normal users, and trolls get sorted behind everyone else.
```

> Implement the `priority_comparison` function that takes two `AccountStatus` arguments and returns `true`, if and only if the first account has a strictly higher priority than the second.

Reviewer 2: ```suggestion
Implement the `has_priority` function that takes two `AccountStatus` arguments and returns `true`, if and only if the first account has a strictly higher priority than the second.
```

Reviewer 2: `has_priority` felt a little more straightforward here, since `priority_comparison` reads to me as if it should return a ranking or other categorization for the respective users.
